test/spectests/spectests.cpp

```diff
@@ -0,0 +1,281 @@
+#include "execute.hpp"
+#include "parser.hpp"
+#include <nlohmann/json.hpp>
+#include <filesystem>
+#include <fstream>
+#include <iostream>
+#include <sstream>
+
+namespace fs = std::filesystem;
+using json = nlohmann::json;
+
+namespace
+{
+constexpr auto JsonExtension = ".json";
+
+template <typename T>
+uint64_t json_to_value(const json& v)
+{
+    return static_cast<uint64_t>(static_cast<T>(std::stoull(v.get<std::string>())));
+}
+
+class test_runner
+{
+public:
+    void run_from_file(const fs::path& path)
+    {
+        std::cout << "Running tests from " << path << "\n";
+
+        std::ifstream test_file{path};
+        json j;
+        test_file >> j;
```

Comment:
Perhaps if there's no non-stream version of this, I'd introduce a tiny helper `json parse_json_from_file(path)` for readability. But this is just a nitpick.